# Chapter 2. Program Structure

## 2.1 Names

Go has 25 keywords
<code>
    break    default     func   interface select
    case     defer       go     map       struct
    chan     else        goto   package   switch
    const    fallthrough if     range     type
    continue for         import return    var
</code>

Predefined names for built-in constants, types, and functions
<code>
Constants:  true false iota nil
[Types](https://golang.org/ref/spec#Types):      int int8 int16 int32 int64
            unit unit8 unit16 uint32 uint64 uintptr
            float32 float64 complex64 complex128
            bool byte rune string error
[Functions](https://golang.org/ref/spec#Built-in_functions):  make len cap new append copy close delete 
            complex real imag 
            panic recover
</code>

In [36]:
package main
import "fmt"

type 整數 int
func 轉換程式(金 整數){
    fmt.Println(金)
}

func main(){
    銅 := 2
    轉換程式(11)
    fmt.Println(銅)
}

main()


11
2


## 2.2 Declarations

In [39]:
//gopl.io/ch2/boiling
// Boiling prints the boiling point of water
package main
import "fmt"

const boilingF = 212.0

func main() {
        var f = boilingF
        var c = (f - 32) * 5 / 9
        fmt.Printf("boiling point = %g°F or %g°C\n", f, c)
        // Output:
        // boiling point = 212°F or 100°C
}

In [40]:
main()

boiling point = 212°F or 100°C


In [41]:
//gopl.io/ch2/ftoc
import "fmt"

func main() {
	const freezingF, boilingF = 32.0, 212.0
	fmt.Printf("%g°F = %g°C\n", freezingF, fToC(freezingF)) // "32°F = 0°C"
	fmt.Printf("%g°F = %g°C\n", boilingF, fToC(boilingF))   // "212°F = 100°C"
}

func fToC(f float64) float64 {
	return (f - 32) * 5 / 9
}

In [42]:
main()

32°F = 0°C
212°F = 100°C


## 2.3 Variables

### 2.3.1 Short Variable Declarations
* Declaration and initialize local variables
  * within functions
  * name := expression
  * type of name is determined by the type of expression
  * A short variable declaration must declare at least one new variable 


In [43]:
import ("image";"image/gif";"math/rand")
nframes := 100
anim := gif.GIF{LoopCount: nframes}
freq := rand.Float64() * 3.0
t := 0.0

In [44]:
i := 100    // an int
var boiling float64 = 100 // a float64
var names []string
var err error
type Point int
var p Point

### 2.3.2 Pointers

In [46]:
x := 1
p := &x           // p, of type *int, points to x
fmt.Println(*p)   // "1"
*p = 21            // equivalent to x = 2
fmt.Println(x)    // "2"

1
21


3 <nil>

In [47]:
var x, y int
fmt.Println(&x==&x, &x==&y, &x== nil) // "true false false"

true false false


17 <nil>

In [48]:
var p = f()
func f() *int {
    v := 1
    return &v
}

fmt.Println(f() == f()) // "false"

false


6 <nil>

In [52]:
func incr(p *int) int {
    *p++ // increments what p points to; does not change p
    return *p
}
v := 1
incr(&v)               // side effect: v is now 2
fmt.Println(incr(&v))  // "3" (and v is 3)

3


2 <nil>

In [1]:
package main
import (
	"flag"
	"fmt"
	"strings"
)

var n = flag.Bool("n", false, "omit trailing newline")
var sep = flag.String("s", " ", "separator")

func main() {
	flag.Parse()
	fmt.Print(strings.Join(flag.Args(), *sep))
	if !*n {
		fmt.Println()
	}
}


In [ ]:
import ("os")
//os.Args = []string{"echo4","-s","/","a","bc","def"};main()
//os.Args = []string{"echo4","-n","a","bc","def"};main()
os.Args = []string{"echo4","-help"}
main()

$ go install gopl.io/ch2/echo4@latest
$ bin/echo4 -help
Usage of ./echo4:
$ bin/echo -s / a bc def
a/bc/def
$ bin/echo -n a bc def
a bc def$

### 2.3.3 The new Function

In [1]:
import "fmt"
//x := 1
//p := &x
p := new(int) //p, of type *int, points to an unnamed int variable
fmt.Println(*p) // "0"
*p = 2 
fmt.Println(*p) // "2"

0
2


2 <nil>

### 2.3.4 Lifetime of Variables

<code>
for t:= 0.0; t < cycles * 2 *math.Pi; i+= res {
    x := math.Sin(t)
    y := math.Sin(t*freq+ phase)
    img.SetColorIndex( size + int(x*size+0.5), size + int(y*size + 0.5), blackIndex)
}
</code>
the variable t is created each time the for loop begins, and variables x and y are created on each iteration of the loop

In [2]:
var global *int
func f() {
    var x int       //x should be heap allocated
    x = 3
    global = &x
}

func g() {
    y := new(int)  //y can be on the stack
    *y = 1
}

//fmt.Println(*global)
f()
fmt.Println(*global)

3


2 <nil>

## 2.4. Assignments

 <code>
    x = 1                      // named variable
    *p = true                  // indirect variable
    person.name = "bob"        // struct field
    count[x] = count[x] * scale // array or slice or map element
 </code>

### 2.4.1. Tuple Assignment

In [5]:
a :=[]int {9,6,4,10,11,12,13}
i, j := 5, 2
x, y := 3, 9

x , y = y, x
a[i], a[j] = a[j], a[i]     //swapping the values
fmt.Println(a)    
fmt.Println(x,y)

[9 6 12 10 11 4 13]
9 3


4 <nil>

In [6]:
func gcd(x,y int) int {
    for y != 0 {
        x, y = y, x%y
    }
    return x
}
func fib(n int) int {
    x, y := 0, 1
    for i := 0; i< n; i++ {
        x, y = y, x+y
    }
    return x
}

In [8]:
fmt.Println(gcd(80,8), gcd(24, 120), gcd(37, 39))
fmt.Println(fib(1), fib(2), fib(3), fib(4), fib(5), fib(6), fib(7), fib(8), fib(9), fib(10))

8 24 1
1 1 2 3 5 8 13 21 34 55


24 <nil>

### 2.4.2 Assignability

In [9]:
modals := []string{"gold","silver","brone"}
fmt.Println(modals)

[gold silver brone]


20 <nil>

In [10]:
modals[0] = "gold"
modals[1] = "silver"
modals[2] = "brone"
fmt.Println(modals)

[gold silver brone]


20 <nil>

## 2.5. Type Declarations

In [14]:
// Copyright © 2016 Alan A. A. Donovan & Brian W. Kernighan.
// License: https://creativecommons.org/licenses/by-nc-sa/4.0/

// See page 39.
//!+
//gopl.io/ch2/tempconv0/celsius.go
// Package tempconv performs Celsius and Fahrenheit temperature computations.
package tempconv

import "fmt"

type Celsius float64
type Fahrenheit float64

const (
	AbsoluteZeroC Celsius = -273.15
	FreezingC     Celsius = 0
	BoilingC      Celsius = 100
)

func CToF(c Celsius) Fahrenheit { return Fahrenheit(c*9/5 + 32) }

func FToC(f Fahrenheit) Celsius { return Celsius((f - 32) * 5 / 9) }

//!-

func (c Celsius) String() string { return fmt.Sprintf("%g°C", c) }

In [15]:
// Copyright © 2016 Alan A. A. Donovan & Brian W. Kernighan.
// License: https://creativecommons.org/licenses/by-nc-sa/4.0/

package tempconv

import "fmt"

func Example_one() {
	{
		//!+arith
		fmt.Printf("%g\n", BoilingC-FreezingC) // "100" °C
		boilingF := CToF(BoilingC)
		fmt.Printf("%g\n", boilingF-CToF(FreezingC)) // "180" °F
		//!-arith
	}
	/*
		//!+arith
		fmt.Printf("%g\n", boilingF-FreezingC)       // compile error: type mismatch
		//!-arith
	*/

	// Output:
	// 100
	// 180
}

func Example_two() {
	//!+printf
	c := FToC(212.0)
	fmt.Println(c.String()) // "100°C"
    fmt.Printf("%v\n", c)   // "100°C"; no need to call String explicitly
    fmt.Printf("%s\n", c)  // "100°C"
	fmt.Println(c)          // "100°C"
	fmt.Printf("%g\n", c)   // "100"; does not call String
	fmt.Println(float64(c)) // "100"; does not call String
	//!-printf

	// Output:
	// 100°C
	// 100°C
	// 100°C
	// 100°C
	// 100
	// 100
}

<code>
$go test gopl.io/ch2/tempconv0
ok      gopl.io/ch2/tempconv0   0.020s
</code>

## 2.6. Packages and Files

In [3]:
// Copyright © 2016 Alan A. A. Donovan & Brian W. Kernighan.
// License: https://creativecommons.org/licenses/by-nc-sa/4.0/

// See page 43.
//!+

// Cf converts its numeric argument to Celsius and Fahrenheit.
package main

import (
	"fmt"
	"os"
	"strconv"

	"gopl.io/ch2/tempconv"
)

func main() {
	for _, arg := range os.Args[1:] {
		t, err := strconv.ParseFloat(arg, 64)
		if err != nil {
			fmt.Fprintf(os.Stderr, "cf: %v\n", err)
			os.Exit(1)
		}
		f := tempconv.Fahrenheit(t)
		c := tempconv.Celsius(t)
		fmt.Printf("%s = %s, %s = %s\n",
			f, tempconv.FToC(f), c, tempconv.CToF(c))
	}
}

//!-

In [4]:
import ("os")
os.Args = []string{"cf","32"}
main()
main()

32°F = 0°C, 32°C = 89.6°F
32°F = 0°C, 32°C = 89.6°F


<code>
$ go install gopl.io/ch2/cf@latest
$./cf 32
32°F = 0°C, 32°C = 89.6°F
$ ./cf 212
212°F = 100°C, 212°C = 413.6°F
$ ./cf -40
-40°F = -40°C, -40°C = -40°F
</code>

### 2.6.2. Package Initialization

In [5]:
var a = b + c    // a initialized third, to 3
var b = f()      // b initialized second, to 2, by callin f
var c = 1        // c initialized first, to 1
func f() int {return c+1}
fmt.Println(a,b,c)

3 2 1


6 <nil>

In [9]:
// Copyright © 2016 Alan A. A. Donovan & Brian W. Kernighan.
// License: https://creativecommons.org/licenses/by-nc-sa/4.0/

// See page 45.

// (Package doc comment intentionally malformed to demonstrate golint.)
//!+
package popcount

// pc[i] is the population count of i.
var pc [256]byte

func init() {
	for i := range pc {
		pc[i] = pc[i/2] + byte(i&1)
	}
}

// PopCount returns the population count (number of set bits) of x.
func PopCount(x uint64) int {
	return int(pc[byte(x>>(0*8))] +
		pc[byte(x>>(1*8))] +
		pc[byte(x>>(2*8))] +
		pc[byte(x>>(3*8))] +
		pc[byte(x>>(4*8))] +
		pc[byte(x>>(5*8))] +
		pc[byte(x>>(6*8))] +
		pc[byte(x>>(7*8))])
}

//!-

In [10]:
// Copyright © 2016 Alan A. A. Donovan & Brian W. Kernighan.
// License: https://creativecommons.org/licenses/by-nc-sa/4.0/

package popcount_test

import (
	"testing"

	"gopl.io/ch2/popcount"
)

// -- Alternative implementations --

func BitCount(x uint64) int {
	// Hacker's Delight, Figure 5-2.
	x = x - ((x >> 1) & 0x5555555555555555)
	x = (x & 0x3333333333333333) + ((x >> 2) & 0x3333333333333333)
	x = (x + (x >> 4)) & 0x0f0f0f0f0f0f0f0f
	x = x + (x >> 8)
	x = x + (x >> 16)
	x = x + (x >> 32)
	return int(x & 0x7f)
}

func PopCountByClearing(x uint64) int {
	n := 0
	for x != 0 {
		x = x & (x - 1) // clear rightmost non-zero bit
		n++
	}
	return n
}

func PopCountByShifting(x uint64) int {
	n := 0
	for i := uint(0); i < 64; i++ {
		if x&(1<<i) != 0 {
			n++
		}
	}
	return n
}

// -- Benchmarks --

func BenchmarkPopCount(b *testing.B) {
	for i := 0; i < b.N; i++ {
		popcount.PopCount(0x1234567890ABCDEF)
	}
}

func BenchmarkBitCount(b *testing.B) {
	for i := 0; i < b.N; i++ {
		BitCount(0x1234567890ABCDEF)
	}
}

func BenchmarkPopCountByClearing(b *testing.B) {
	for i := 0; i < b.N; i++ {
		PopCountByClearing(0x1234567890ABCDEF)
	}
}

func BenchmarkPopCountByShifting(b *testing.B) {
	for i := 0; i < b.N; i++ {
		PopCountByShifting(0x1234567890ABCDEF)
	}
}

// Go 1.6, 2.67GHz Xeon
// $ go test -cpu=4 -bench=. gopl.io/ch2/popcount
// BenchmarkPopCount-4                  200000000         6.30 ns/op
// BenchmarkBitCount-4                  300000000         4.15 ns/op
// BenchmarkPopCountByClearing-4        30000000         45.2 ns/op
// BenchmarkPopCountByShifting-4        10000000        153 ns/op
//
// Go 1.6, 2.5GHz Intel Core i5
// $ go test -cpu=4 -bench=. gopl.io/ch2/popcount
// BenchmarkPopCount-4                  200000000         7.52 ns/op
// BenchmarkBitCount-4                  500000000         3.36 ns/op
// BenchmarkPopCountByClearing-4        50000000         34.3 ns/op
// BenchmarkPopCountByShifting-4        20000000        108 ns/op
//
// Go 1.7, 3.5GHz Xeon
// $ go test -cpu=4 -bench=. gopl.io/ch2/popcount
// BenchmarkPopCount-12                 2000000000        0.28 ns/op
// BenchmarkBitCount-12                 2000000000        0.27 ns/op
// BenchmarkPopCountByClearing-12       100000000        18.5 ns/op
// BenchmarkPopCountByShifting-12       20000000         70.1 ns/op

ERROR: repl.go:37:12: mismatched types in binary operation & between <uint64> and <int>: x & (1 << i)

<code>
$ go test -cpu=2 -bench=. gopl.io/ch2/popcount
goos: linux
goarch: amd64
pkg: gopl.io/ch2/popcount
BenchmarkPopCount-2                     1000000000               0.704 ns/op
BenchmarkBitCount-2                     1000000000               0.361 ns/op
BenchmarkPopCountByClearing-2           63276454                18.3 ns/op
BenchmarkPopCountByShifting-2           21629844                56.6 ns/op
PASS
ok      gopl.io/ch2/popcount    3.661s
</code>

## 2.7. Scope

In [11]:
func f() {}
var g = "g"

func main(){
    f := "f"
    fmt.Println(f) // "f"; local var f shadows package-level func f
    fmt.Println(g) // "g"; package-level var
    //fmt.Println(h) // compile error: undefined: h
}

In [12]:
main()

f
g


In [13]:
// three different variables called x for each declaration appears in a different lexical block
func main(){
    x := "hello!"
    for i:= 0; i < len(x); i++ {
        x := x[i]
        if x != '!' {
            x := x + 'A' - 'a'
            fmt.Printf("%c", x)   // "HELLO" (one letter per iteration)
        }
    }
}

In [14]:
main()

HELLO

In [15]:
func main(){
    x := "hello"
    for _, x := range x {
        x := x + 'A' - 'a' + 2
        fmt.Printf("%c", x)  // "HELLO" (one letter per iteration)
    }
}

In [16]:
main()

JGNNQ

In [17]:
func f() int { return 1}
func g(x int) int { return 1}

if x:= f(); x == 0 {
    fmt.Println(x)
} else if y:= g(x); x == y {
    fmt.Println(x,y)
}
//fmt.Println(x, y) // compile error: x and y are not visible here

1 1


In [18]:
import "os"
    fname := "dup.txt"
    if f, err := os.Open(fname); err != nil {
        //return err
    }
    f.Stat()
    f.Close()

ERROR: repl.go:6:5: not a package: "f" in f.Stat <*ast.SelectorExpr>

In [19]:
f, err := os.Open("dup.txt")
if err != nil {
    //return err
}
f.Stat()
f.Close()

In [21]:
if f, err := os.Open(fname); err != nil {
    //return err
} else {
    // f and err are visible here too
    f.Stat()
    f.Close()
}

In [22]:
import ( "os"; "log")
var cwd string
func init() {
    cwd, err := os.Getwd() // compile error: unused: cwd
    if err != nil {
        log.Fatalf("os.Getwd failed: %v", err)
    }
}

In [23]:
var cwd string
func init() {
    cwd, err := os.Getwd()
    if err != nil {
        log.Fatalf("os.Getwd failed: %v", err) // log output obfuscated the bug
    }
}

In [24]:
var cwd string    // avoid := by declaring err in a separte var declaration
func init(){
    var err error
    cwd, err = os.Getwd()
    if err != nil {
        log.Fatalf("os.Getwd failed: %v", err)
    }
}